rasa/core/policies/ensemble.py

```diff
@@ -91,590 +220,318 @@ def _check_for_important_policies(self) -> None:
-
-        if tracker.latest_action_name == ACTION_LISTEN_NAME:
-            if winning_prediction.is_end_to_end_prediction:
-                logger.debug("Made e2e prediction using user text.")
-                logger.debug("Added `DefinePrevUserUtteredFeaturization(True)` event.")
-                winning_prediction.events.append(
-                    DefinePrevUserUtteredFeaturization(True)
-                )
-            else:
-                logger.debug("Made prediction using user intent.")
-                logger.debug("Added `DefinePrevUserUtteredFeaturization(False)` event.")
-                winning_prediction.events.append(
-                    DefinePrevUserUtteredFeaturization(False)
-                )
-
-        logger.debug(f"Predicted next action using {winning_prediction.policy_name}.")
-        return winning_prediction
-
-
-def _check_policy_for_forms_available(
-    domain: Domain, ensemble: Optional["PolicyEnsemble"]
-) -> None:
-    if not ensemble:
-        return
-
-    has_policy_for_forms = ensemble is not None and any(
-        isinstance(policy, RulePolicy) for policy in ensemble.policies
-    )
-
-    if domain.form_names and not has_policy_for_forms:
-        raise InvalidDomain(
-            "You have defined a form action, but have not added the "
-            f"'{RulePolicy.__name__}' to your policy ensemble. Either "
-            f"remove all forms from your domain or add the '{RulePolicy.__name__}' "
-            f"to your policy configuration."
-        )
+        if best_idx < 0 or best_idx >= len(predictions):
+            raise ValueError(
+                f"Expected given index to be pointing towards the best prediction "
+                f"among the given predictions. Received index {best_idx} and a list "
+                f"of predictions of length {len(predictions)}"
+            )
 
+        # combine all mandatory events in the given order
+        events = [event for prediction in predictions for event in prediction.events]
 
-class InvalidPolicyConfig(RasaException):
-    """Exception that can be raised when policy config is not valid."""
+        # append optional events from the "best prediction"
+        best = predictions[best_idx]
+        events += best.optional_events
 
-    pass
+        # if the sequence of events ends with a user utterance, append an event
+        # defining the featurization
+        if tracker.latest_action_name == ACTION_LISTEN_NAME:
+            was_end_to_end_prediction = best.is_end_to_end_prediction
+            if was_end_to_end_prediction:
+                logger.debug("Made prediction using user text.")
+            else:
+                logger.debug("Made prediction without user text")
+            event = DefinePrevUserUtteredFeaturization(was_end_to_end_prediction)
+            events.append(event)
+            logger.debug(f"Added `{event}` event.")
+        return events
```

Comment:
Can we somehow move this right into the policy? Or would this be more messy?

Reply:
Oh, I see - "choose_events" is a poor name for this as it is adding additional events maybe and we could pull the  `DefinePrevUserUtteredFeaturization` out .... Or, do you mean the whole "_choose_events"?